dll/shellext/cabview/cabview.h

```diff
@@ -0,0 +1,279 @@
+/*
+ * PROJECT:     ReactOS CabView Shell Extension
+ * LICENSE:     GPL-2.0-or-later (https://spdx.org/licenses/GPL-2.0-or-later)
+ * PURPOSE:     Main header file
+ * COPYRIGHT:   Copyright 2024 Whindmar Saksit <whindsaks@proton.me>
+ */
+
+#pragma once
+#include "precomp.h"
+#include "resource.h"
+
+#define FLATFOLDER TRUE
+
+EXTERN_C const GUID CLSID_CabFolder;
+extern HINSTANCE g_hInst;
+
+enum EXTRACTCALLBACKMSG { ECM_BEGIN, ECM_FILE, ECM_PREPAREPATH, ECM_ERROR };
+struct EXTRACTCALLBACKDATA
+{
+    LPCWSTR Path;
+    const FDINOTIFICATION *pfdin;
+    HRESULT hr;
+};
+typedef HRESULT (CALLBACK*EXTRACTCALLBACK)(EXTRACTCALLBACKMSG msg, const EXTRACTCALLBACKDATA &data, LPVOID cookie);
+HRESULT ExtractCabinet(LPCWSTR cab, LPCWSTR destination, EXTRACTCALLBACK callback, LPVOID cookie);
+
+
+#include <pshpack1.h>
+struct CABITEM
+{
+    WORD cb;
+    WORD Unknown; // Not sure what Windows uses this for, we always store 0
+    UINT Size;
+    WORD Date, Time; // DOS
+    WORD Attrib;
+    WORD NameOffset;
+    WCHAR Path[ANYSIZE_ARRAY];
+
+#if FLATFOLDER
+    inline bool IsFolder() const { return false; }
+#else
+    inline BOOL IsFolder() const { return Attrib & FILE_ATTRIBUTE_DIRECTORY; }
+#endif
+    enum { FSATTS = FILE_ATTRIBUTE_READONLY | FILE_ATTRIBUTE_HIDDEN | FILE_ATTRIBUTE_SYSTEM |
+                    FILE_ATTRIBUTE_ARCHIVE | FILE_ATTRIBUTE_DIRECTORY };
+    WORD GetFSAttributes() const { return Attrib & FSATTS; }
+    LPCWSTR GetName() const { return Path + NameOffset; }
+
+    template<class PIDL> static CABITEM* Validate(PIDL pidl)
+    {
+        CABITEM *p = (CABITEM*)pidl;
+        return p && p->cb > FIELD_OFFSET(CABITEM, Path[1]) && p->Unknown == 0 ? p : NULL;
+    }
+};
+#include <poppack.h>
+
+class CEnumIDList :
+    public CComObjectRootEx<CComMultiThreadModelNoCS>,
+    public IEnumIDList
+{
+protected:
+    HDPA m_Items;
+    ULONG m_Pos;
+
+public:
+    static int CALLBACK DPADestroyCallback(void *pidl, void *pData)
+    {
+        SHFree(pidl);
+        return TRUE;
+    }
+
+    CEnumIDList() : m_Pos(0)
+    {
+        m_Items = DPA_Create(0);
+    }
+
+    virtual ~CEnumIDList()
+    {
+        DPA_DestroyCallback(m_Items, DPADestroyCallback, NULL);
+    }
+
+    int FindNamedItem(PCUITEMID_CHILD pidl) const;
+    HRESULT Fill(LPCWSTR path, HWND hwnd = NULL, SHCONTF contf = 0);
+    HRESULT Fill(PCIDLIST_ABSOLUTE pidl, HWND hwnd = NULL, SHCONTF contf = 0);
+
+    HRESULT Append(LPCITEMIDLIST pidl)
+    {
+        return DPA_AppendPtr(m_Items, (void*)pidl) != DPA_ERR ? S_OK : E_OUTOFMEMORY;
+    }
+
+    UINT GetCount() const { return m_Items ? DPA_GetPtrCount(m_Items) : 0; }
+
+    // IEnumIDList
+    IFACEMETHODIMP Next(ULONG celt, PITEMID_CHILD *rgelt, ULONG *pceltFetched)
+    {
+        if (!rgelt)
+            return E_INVALIDARG;
+        HRESULT hr = S_FALSE;
+        UINT count = GetCount(), fetched = 0;
+        if (m_Pos < count && fetched < celt)
```

Comment:
It can be a loop yes but just returning one item per call is also allowed (DefView only asks for one per call so there is no performance gain to be had here).